File: github_pr.py
import requests
from attachments import create_attachments
from collections import defaultdict
from dateutil.parser import parse

class GithubPR:

    def initialize(self, link):
        link_parts = link.split('/')
        pull_number = link_parts[len(link_parts)-1]
        repo = link_parts[len(link_parts)-3]
        owner = link_parts[len(link_parts)-4]

        pr_request = 'https://api.github.com/repos/' + owner + '/' + repo + '/' + 'pulls/' + pull_number
        result = requests.get(pr_request).json()

        self.title = result['title']
        self.user_login = result['user']['login']
        self.user_avatar = result['user']['avatar_url']
        self.labels = []
        for label in result['labels']:
            self.labels.append({'name': label['name'], 'color': label['color']})
        
        self.updated_at = result["updated_at"]
        self.state = result['state']
            
        pr_request_reviewers = pr_request + '/requested_reviewers'
        pr_request_reviews = pr_request + '/reviews'
        
        results = requests.get(pr_request_reviewers).json()

        self.reviewers = defaultdict()
        for result in results['users'] + results['teams']:
            # self.reviewers.append({
            #     'login': result['login'],
            #     'avatar': result['avatar_url']
            # })
            self.reviewers[result['login']] = {
                'avatar': result['avatar_url'],
                'state': "UNDER_REVIEW",
                'submitted': None
            }

        results = requests.get(pr_request_reviews).json()
        for result in results:
            login = result['user']['login']
            if not self.reviewers[login]['submitted']:
                self.reviewers[login] = {
                    'avatar': result['user']['avatar_url'],
                    'state': result['state'],
                    'submitted': result['submitted_at']
                }
            else:
                curr_date = parse(self.reviewers[login]['submitted'])
                inc_date = parse(result['submitted_at'])

                if inc_date > curr_date:
                    self.reviewers[login] = {
                        'avatar': result['user']['avatar_url'],
                        'state': result['state'],
                        'submitted': result['submitted_at']
                    }
        
        count = 0
        for key in self.reviewers.keys():
            if self.reviewers[key]['state'] == 'APPROVED':
                count += 1
        if count >= 2:
            self.color = "#00FF00"
        else:
            self.color = '#FFA500'
            
    def create_pr_message(self):
        attachments = create_attachments(
            self.title, self.labels, self.user_avatar, self.user_login,
            self.updated_at, self.state, self.reviewers, self.color)
        blocks = [
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": "A PR has been raised"
                }
            }
	    ]
        return attachments, blocks

File: amplify/backend/function/prmatelayer51a84423/opt/github_pr.js
const { createAttachments } = require('./attachments');
const { Octokit } = require('@octokit/rest')
const octokit = new Octokit({ auth: "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"})

class GithubPR {

    initialize = async (link) => {
        const linkParts = link.split('/');
        this.pull_number = linkParts[linkParts.length-1];
        this.repo = linkParts[linkParts.length-3];
        this.owner = linkParts[linkParts.length-4];

        try {
            let response = await octokit.request('GET /repos/:owner/:repo/pulls/:num', {
                owner: this.owner,
                repo: this.repo,
                num: this.pull_number
            });
            console.log('res: ', response)
            let result = response['data'];
            this.title = result['title'];
            this.user_login = result['user']['login'];
            this.user_avatar = result['user']['avatar_url'];
            this.labels = [];
            for (const label of result['labels']) {
                this.labels.push({'name': label['name'], 'color': label['color']});
            }
            this.updated_at = result["updated_at"];
            this.state = result['state'];

            this.reviewers = {};
            response = await octokit.request('GET /repos/:owner/:repo/pulls/:num/requested_reviewers', {
                owner: this.owner,
                repo: this.repo,
                num: this.pull_number
            });
            let results = response['data'];
            const reviewers = results['users'].concat(results['teams']);

            reviewers.forEach(result => {
                this.reviewers[result['login']] = {
                    'avatar': result['avatar_url'],
                    'state': "UNDER_REVIEW",
                    'submitted': null
                };
            });

            response = await octokit.request('GET /repos/:owner/:repo/pulls/:num/reviews', {
                owner: this.owner,
                repo: this.repo,
                num: this.pull_number
            });
            results = response['data'];
            results.forEach(result => {
                const login = result['user']['login'];
                if (login !== this.user_login) {
                    if (!(login in this.reviewers)) {
                        this.reviewers[login] = {
                            'avatar': result['user']['avatar_url'],
                            'state': result['state'],
                            'submitted': result['submitted_at']
                        }
                    }
                    else if (!this.reviewers[login]['submitted']) {
                        this.reviewers[login] = {
                            'avatar': result['user']['avatar_url'],
                            'state': result['state'],
                            'submitted': result['submitted_at']
                        };
                    }
                    else {
                        const curr_date = new Date(this.reviewers[login]['submitted']);
                        const inc_date = new Date(result['submitted_at']);
                        
                        if (inc_date > curr_date) {
                            this.reviewers[login] = {
                                'avatar': result['user']['avatar_url'],
                                'state': result['state'],
                                'submitted': result['submitted_at']
                            };
                        }
                    }
                }
            });

            let count = 0;
            Object.keys(this.reviewers).forEach(key => {
                if (this.reviewers[key]['state'] == 'APPROVED') {
                    count += 1;
                }
            });
            if (count >= 2) {
                this.color = "#00FF00";
            }
            else {
                this.color = '#FFA500';
            }
            if (this.state === 'closed') {
                this.color = '#6d00c1'
            }
        }
        catch(err) {
            console.log(err);
        }
    }

    createPrMessage = () => {
        const attachments = createAttachments(
            this.title, this.labels, this.user_avatar, this.user_login,
            this.updated_at, this.state, this.reviewers, this.color);
        const blocks = [
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": `A PR has been raised by <@${this.user_login}>`
                }
            }
        ];
        return { attachments, blocks };
    }
}

module.exports = {
    GithubPR
};